src/ResolvedPeers.h

```diff
@@ -110,6 +106,24 @@ class ResolvedPeers: public RefCountable
     size_type availablePaths = 0;
 };
 
+
+class ResolvedPeerPath
+{
+public:
+    explicit ResolvedPeerPath(const Comm::ConnectionPointer &conn, const ResolvedPeers::size_type pos) : available(true), connection_(conn), position_(pos) {}
+    ResolvedPeerPath() : available(false), connection_(nullptr), position_(ResolvedPeers::npos) {}
+
+    Comm::ConnectionPointer connection() const { return connection_; }
+    ResolvedPeers::size_type position() const { return position_; }
+    explicit operator bool() const { return static_cast<bool>(connection_); }
+
+    bool available; ///< whether this path may be used (i.e., has not been tried already)
+
+private:
+    Comm::ConnectionPointer connection_; ///< (the address of) a path
+    ResolvedPeers::size_type position_; ///< the index of the path
```

Comment:
It feels rather odd to store the "position of x in y" in y itself.

* If that was an accident, let's revert that part of the change. It _is_ a good idea to add something like a `ResolvedCandidate` class that would pair the connection and the position, but it would only be stored _outside_ of the ResolvedPeers class.
* Otherwise, let's discuss the need for this oddity.

Reply:
After looking at HappyConnOpener code, I still think we should add a wrapper class to hold a connection pointer and its (private) ResolvedPeers position. Moreover, I think you can minimize visible/conceptual changes by treating this new class as a smart Connection pointer of sorts -- add `*` and `->` operators that would return a Comm::Connection reference and a Comm::Connection pointer, respectively (just like an existing Comm::ConnectionPointer would). The already discussed `bool` operator will return whether the stored connection pointer is set.

If you the `*` operator returns a reference to the object itself (instead of Connection), then even fewer code changes may be required, but I do not think it is a good idea to de-synchronize  `*` and `->` operators like that.

I would call the new class ResolvedPeer.